# Introduction to DPLYR

# A.1 Working with data in R

### A.1.1 What is a Data Frame? 
- Think of it as an excel sheet with data
- In many cases:
    - Rows are observations (e.g. people, households, countries, time)
    - Columns are variables (e.g. GDP, life expectancy)
        
### A.1.2 Why don't we just use excel if Data Frames have similar structures?
- For many reasons:
    - We can easily copy dataframes and mainpulate
    - We have access to multiple dataframes with only one line of code
    - R is free!

### A.1.3 What are the other advantages of using R and dataframes?
- We can be lazy and use the thousands of free libraries to easily:
    - Easily manipulate data (Today's topic)
    - Download data directly from the internet
    - Viusualize our data (graphing etc.)
    - Build models (Regression, Machine learning, Neural Networks)
    
    
- You have all used libraries before, perhaps without knowing it!
    - This is done in R in two steps: 
        1. install.packages("Package name") Downloads package
        2. library(Package name) Imports package

### A.1.4 What is the best way to manipulate Data Frames?
#### In R: DPLYR!
In python: Pandas

# 1. Basics

## 1.1 Find file [path](https://automatetheboringstuff.com/2e/chapter9/)

![title](Images/File_path.png)

In [1]:
path = getwd() #Sets current working directory
print(paste("The current working directory is: ", getwd(), sep=""))

[1] "The current working directory is: /Users/corybaird/Desktop/Tutorials/R/W1_DPLYR"


## 1.2 Import files
### 1.2.1 Import from: current working directory



#### 1.2.1.1 Lets import that file from an internal data set!


In [8]:
#Download internal data set: cars data
data("mtcars") 

#Show the first two lines of the data set
head(mtcars, 2)

mpg cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4   
Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4

##### 1.2.1.1.1 Write the internal data set to csv file

In [ ]:
write.csv(mtcars, "cars.csv") #Name the dataset "cars.csv" and save to current working directory

#### 1.2.1.2 Now import "cars.csv" which is now in our current directory

In [3]:
# read.csv("cars.csv")
df = read.csv("cars.csv") #We often name our dataframes df_NAME

##### 1.2.1.2.1 Check the first two lines of the cars.csv file now imported as "df"

In [5]:
head(df,2)

X             mpg cyl disp hp  drat wt    qsec  vs am gear carb
1 Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4   
2 Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4

### 1.2.2  Import from: not in our current directory

We need to know the file path of the file!!

#### 1.2.2.1 Create new directory (folder) and place new data set

In [6]:
dir.create('New_directory') #Creates new folder

setwd("New_directory") #Changes working directory

data("USArrests") #Imports internal data set

write.csv(USArrests, "arrests.csv") #Name the dataset "arrests.csv" and save to current working directory

setwd(path) #Returns to our original file path!!!

In [7]:
getwd() #Checks our path which is not in the folder we just created

[1] "/Users/corybaird/Desktop/Tutorials/R/W1_DPLYR"

#### 1.2.2.2 Import "arrests.csv" from the "New_directory" folder

In [6]:
df_arrests = read.csv("New_directory/arrests.csv")

In [7]:
head(df_arrests,2)

X       Murder Assault UrbanPop Rape
1 Alabama 13.2   236     58       21.2
2 Alaska  10.0   263     48       44.5

# 2. DPLYR

We will use cars data set we named df

## 2.1 Basics

### 2.1.1 Download and import dplyr package

In [8]:
#install.packages("dplyr") #Already installed!
library(dplyr)

Warning message:
“package ‘dplyr’ was built under R version 3.6.3”
Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



### 2.1.2 Convert Data Frame into tibble (DPLYR's special Data Frame)

In [9]:
dp_df = as_tibble(df)

### 2.2 Select columns

In [10]:
#We can list all the column names from a Data Frame
colnames(dp_df)

[1] "X"    "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"  
[11] "gear" "carb"

### 2.2.1 Select one column

In [11]:
dp_df %>% #Dataframe
select(X)%>% #Select column named X
head(3) #Shows only first 3 obs

X            
1 Mazda RX4    
2 Mazda RX4 Wag
3 Datsun 710

### 2.2.2 Select multiple columns

In [12]:
dp_df %>% #Dataframe
select(cyl, disp, hp, wt)%>% #Select column named X
head(2) #Shows only first 2 obs

cyl disp hp  wt   
1 6   160  110 2.620
2 6   160  110 2.875

### 2.2.3 QUESTION: Select columns and create new Data Frame

### 2.2.4 Select by column name

#### 2.2.4.1 Select by name that starts with h

In [13]:
dp_df %>% #Dataframe
select(starts_with('h'))%>% #Select column named X
head(2) #Shows only first 2 obs

hp 
1 110
2 110

#### 2.2.4.2 Select by name that starts with t

In [14]:
dp_df %>% #Dataframe
select(ends_with('t'))%>% #Select column named X
head(2) #Shows only first 2 obs

drat wt   
1 3.9  2.620
2 3.9  2.875

#### 2.2.4.3 Select by name that contains c

In [15]:
dp_df %>% #Dataframe
select(contains('c'))%>% #Select column named X
head(2) #Shows only first 2 obs

cyl qsec  carb
1 6   16.46 4   
2 6   17.02 4

### 2.2.4 Pull: similar to select expect it only extracts values

In [16]:
dp_df %>% #Dataframe
pull(var = cyl)

[1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4

## 2.3 Filter 

In [17]:
head(dp_df,2)

X             mpg cyl disp hp  drat wt    qsec  vs am gear carb
1 Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4   
2 Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4

### 2.3.1 Filter by 1 condition

In [18]:
dp_df %>% 
filter(mpg>22)%>% #filter all observations over 25 mpg
head(2)

X           mpg  cyl disp  hp drat wt    qsec  vs am gear carb
1 Datsun 710  22.8 4   108.0 93 3.85 2.320 18.61 1  1  4    1   
2 Merc 240D   24.4 4   146.7 62 3.69 3.190 20.00 1  0  4    2   
3 Merc 230    22.8 4   140.8 95 3.92 3.150 22.90 1  0  4    2   
4 Fiat 128    32.4 4    78.7 66 4.08 2.200 19.47 1  1  4    1   
5 Honda Civic 30.4 4    75.7 52 4.93 1.615 18.52 1  1  4    2

### 2.3.2 Filter by 2 conditions

In [19]:
dp_df %>% 
filter(mpg>22 & mpg<35) %>% head(2)

X           mpg  cyl disp  hp drat wt    qsec  vs am gear carb
1 Datsun 710  22.8 4   108.0 93 3.85 2.320 18.61 1  1  4    1   
2 Merc 240D   24.4 4   146.7 62 3.69 3.190 20.00 1  0  4    2   
3 Merc 230    22.8 4   140.8 95 3.92 3.150 22.90 1  0  4    2   
4 Fiat 128    32.4 4    78.7 66 4.08 2.200 19.47 1  1  4    1   
5 Honda Civic 30.4 4    75.7 52 4.93 1.615 18.52 1  1  4    2

### 2.3.3 Filter by gear not equal to 4

In [20]:
dp_df %>% 
filter(gear!=4)%>% head(2)

X                 mpg  cyl disp hp  drat wt    qsec  vs am gear carb
1 Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
2 Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2

### 2.3.4 QUESTION: Filter by vs=1

### 2.3.5 Filter by Mazda 
-grepl function

In [21]:
dp_df %>% filter(grepl("Maz",X))

X             mpg cyl disp hp  drat wt    qsec  vs am gear carb
1 Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4   
2 Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4

## 2.4 Arrange

### 2.4.1 Arrange by column (Ascending)

In [22]:
dp_df %>% 
arrange(mpg)%>% 
head(3)

X                   mpg  cyl disp hp  drat wt    qsec  vs am gear carb
1 Cadillac Fleetwood  10.4 8   472  205 2.93 5.250 17.98 0  0  3    4   
2 Lincoln Continental 10.4 8   460  215 3.00 5.424 17.82 0  0  3    4   
3 Camaro Z28          13.3 8   350  245 3.73 3.840 15.41 0  0  3    4

### 2.4.2 Arrange  (Descending)

In [23]:
dp_df %>% 
arrange(desc(mpg))%>% 
head(3)

X              mpg  cyl disp hp drat wt    qsec  vs am gear carb
1 Toyota Corolla 33.9 4   71.1 65 4.22 1.835 19.90 1  1  4    1   
2 Fiat 128       32.4 4   78.7 66 4.08 2.200 19.47 1  1  4    1   
3 Honda Civic    30.4 4   75.7 52 4.93 1.615 18.52 1  1  4    2

## 2.5 Mutate (Create new column)

### 2.5.1 Mutate

In [24]:
dp_df %>% 
mutate(mpg_10 = mpg/10) %>% head(2)

X             mpg cyl disp hp  drat wt    qsec  vs am gear carb mpg_10
1 Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4    2.1   
2 Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4    2.1

#### 2.5.1.1 Mutate: Various functions
1. Cumulative sum: cumsum(X)
2. Lag: lag(X)
3. percent_rank(X)

In [25]:
dp_df %>% 
mutate(cumulative_wt = cumsum(wt), #cumulative
       lag_wt = lag(wt), #Opposite function: lead()
       Percent_wt = percent_rank(wt)
      ) %>% head(2)

X             mpg cyl disp hp  drat wt    qsec  vs am gear carb cumulative_wt
1 Mazda RX4     21  6   160  110 3.9  2.620 16.46 0  1  4    4    2.620        
2 Mazda RX4 Wag 21  6   160  110 3.9  2.875 17.02 0  1  4    4    5.495        
  lag_wt Percent_wt
1   NA   0.2580645 
2 2.62   0.3548387

### 2.5.2 Transmute
- Same as mutate expect it drops other columns

In [26]:
dp_df %>% 
transmute(mpg_10 = mpg/10)%>% head(2)

mpg_10
1 2.1   
2 2.1

### 2.5.3 Mutate_all

- Mutate all selected columns by certain function
- This example shows log
    - funs(log(.))

In [27]:
dp_df[, 2:5]%>% 
mutate_all(funs(log(.)))%>%
head(3)

Warning message:
“funs() is soft deprecated as of dplyr 0.8.0
Please use a list of either functions or lambdas: 

  # Simple named list: 
  list(mean = mean, median = median)

  # Auto named with `tibble::lst()`: 
  tibble::lst(mean, median)

  # Using lambdas
  list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))
This warning is displayed once per session.”

mpg      cyl      disp     hp      
1 3.044522 1.791759 5.075174 4.700480
2 3.044522 1.791759 5.075174 4.700480
3 3.126761 1.386294 4.682131 4.532599

### 2.5.4 Mutate_at

Mutates only at specified columns

In [28]:
dp_df%>% 
mutate_at(vars(mpg, disp, hp), funs(log(.)))%>% 
head(2)

X             mpg      cyl disp     hp      drat wt    qsec  vs am gear carb
1 Mazda RX4     3.044522 6   5.075174 4.70048 3.9  2.620 16.46 0  1  4    4   
2 Mazda RX4 Wag 3.044522 6   5.075174 4.70048 3.9  2.875 17.02 0  1  4    4

## 2.6 Summarise

In [29]:
dp_df %>% 
summarise(mean_mpg = mean(mpg),
        median_mpg = median(mpg),
         sum_mpg = sum(mpg),
         count_mpg = n(),
         first_obs = first(mpg),
         last_obs = last(mpg),
         variance = var(mpg),
         sd = sd(mpg))

mean_mpg median_mpg sum_mpg count_mpg first_obs last_obs variance sd      
1 20.09062 19.2       642.9   32        21        21.4     36.3241  6.026948

## 2.7 Groupby with summarise

In [30]:
dp_df %>% 
group_by(gear) %>% 
summarise(mpg_by_gear = mean(mpg))

gear mpg_by_gear
1 3    16.10667   
2 4    24.53333   
3 5    21.38000

# 3. Other import DPLYR functions

## 3.1 Count

### 3.1.1 Count number of observations by gear

In [31]:
dp_df %>% 
count(gear)

gear n 
1 3    15
2 4    12
3 5     5

## 3.2 Distinct

In [32]:
dp_df %>% 
distinct(X)%>% head(3)

X            
1 Mazda RX4    
2 Mazda RX4 Wag
3 Datsun 710

## 3.3 Sample data

Used for train and test split in ML and NN

### 3.3.1 Sample 50% of the data

In [33]:
dp_df %>% 
sample_frac(.5)%>% head(3)

X             mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1 Duster 360    14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
2 Toyota Corona 21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
3 Honda Civic   30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2

### 3.3.2 Sample 10 observations

In [34]:
dp_df %>% 
sample_n(10)%>% head(3)

X                   mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1 Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
2 Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
3 Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3

## 3.4 Slice
Pick out rows by observation number

In [35]:
dp_df %>% 
slice(10:12)

X          mpg  cyl disp  hp  drat wt   qsec vs am gear carb
1 Merc 280   19.2 6   167.6 123 3.92 3.44 18.3 1  0  4    4   
2 Merc 280C  17.8 6   167.6 123 3.92 3.44 18.9 1  0  4    4   
3 Merc 450SE 16.4 8   275.8 180 3.07 4.07 17.4 0  0  3    3

## 3.5 add_count

In [36]:
dp_df %>% 
add_count()%>% 
head(3)

X             mpg  cyl disp hp  drat wt    qsec  vs am gear carb n 
1 Mazda RX4     21.0 6   160  110 3.90 2.620 16.46 0  1  4    4    32
2 Mazda RX4 Wag 21.0 6   160  110 3.90 2.875 17.02 0  1  4    4    32
3 Datsun 710    22.8 4   108   93 3.85 2.320 18.61 1  1  4    1    32

# Questions

In [37]:
head(dp_df, 10)

X                 mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
1  Mazda RX4         21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
2  Mazda RX4 Wag     21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
3  Datsun 710        22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
4  Hornet 4 Drive    21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
5  Hornet Sportabout 18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
6  Valiant           18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
7  Duster 360        14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
8  Merc 240D         24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
9  Merc 230          22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
10 Merc 280          19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4

## Q.1. Show mean mpg by gear

## Q.2. Show all observations with: am=1 & vs=0

## Q.3 Create new column for Kilometers per gallon (mulitply mpg*1.6)

## Q.4 Show all observations with a gear = 4 but instead of showing the whole frame only showing cyl and disp columns